File: Top 50 Coding Interview Questions/10-How is the merge sort algorithm implemented/MergeSort.c
#include<stdio.h>

void mergesort(int[], int, int);
void merge(int[], int, int, int);

void display(int arr[],int size){
    int i;
    for(i = 0;i < size; i++){
        printf("%d ",arr[i]);
    }
    printf("\n");
}

void main(){
    int a[10]={
        11,6,8,9,2,3,54,67,89,20
    };
    int i;
    int size = sizeof(a)/sizeof(a[0]);
    display(a,size);


    mergesort(a,0,size-1);
    display(a,size);
}

void mergesort(int a[],int left, int right){
    int mid;
    if(left < right){
        //can also use mid = left +(right - left)/2
        //this can avoid data type overflow
        mid = (left + right)/2;

        //recursive calls to sor first half and second alf subarray
        mergesort(a,left,mid);
        mergesort(a,mid+1,right);
        merge(a,left,mid,right);
    }
}

void merge(int arr[], int left, int mid, int right){
    int i,j,k;
    int n1 = mid -left +1;
    int n2 = right - mid;

    //reate temp array to stor left and right sub arrays
    int L[n1], R[n2];

    //copying data to temp arrays L[] and R[]
    for(i = 0; i <n1;i++)
      L[i] = arr[left + i];
    for(j = 0; j <n2 ; j++)
      R[j] = arr[mid +1 +j];

    //here we merge the temp arrays back into arr[l..r]
    i = 0; //starting index of L[i]
    j = 0;//Starting index of R[i]
    k = left ;//starting index of merged subarray

    while(i < n1 && j < n2){
        //place the smaller item at arr[k] pos
        if(L[i] <= R[j]){
            arr[k] = L[i];
            i++;
        }
        else{
            arr[k] = R[j];
            j++;
        }
        k++;
    }
     //copy the remaining elements of L[], if any
     while(i < n1){
        arr[k] = L[i];
        i++;
        k++;
     }
     //copy the remaining elements of R[], if any
     while(j < n2){
        arr[k] = R[j];
        j++;
        k++;
     }
}

File: Top 50 Coding Interview Questions/10-How is the merge sort algorithm implemented/MergeSort.py
# Function to merge two subarrays
def merge(arr, left, mid, right):
    # Sizes of the two subarrays
    n1 = mid - left + 1
    n2 = right - mid

    # Create temporary arrays
    L = [arr[left + i] for i in range(n1)]
    R = [arr[mid + 1 + i] for i in range(n2)]

    # Merge temporary arrays back into arr
    i = 0  # Index for L
    j = 0  # Index for R
    k = left  # Index for main array

    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    # Copy any remaining elements of L[]
    for x in range(i, n1):
        arr[k] = L[x]
        k += 1

    # Copy any remaining elements of R[]
    for x in range(j, n2):
        arr[k] = R[x]
        k += 1

# Function to perform merge sort
def merge_sort(arr, left, right):
    if left < right:
        # Find the middle point
        mid = left + (right - left) // 2

        # Recursively sort first and second halves
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)

        # Merge the sorted halves
        merge(arr, left, mid, right)

# Test the Merge Sort
arr = [38, 27, 43, 3, 9, 82, 10]
print("Original array:", arr)

merge_sort(arr, 0, len(arr) - 1)

print("Sorted array:", arr)
